### PR TITLE
Fix pipeline compilation to make it fail when multiple codecs per plugin are specified

### LINKED ISSUE #14773: Logstash should provide better error messages when multiple codecs are specified
If a plugin is erroneously configured with multiple codecs, the error message is very unhelpful:

```
input {
  stdin {
    codec => plain{
      charset => "UTF-8"
    } 
    codec => plain 
  }
}

output {
  stdout { codec => rubydebug }
}
```

Results in the following error:

```
[2022-11-22T10:13:22,964][ERROR][logstash.agent           ] Failed to execute action {:action=>LogStash::PipelineAction::Create/pipeline_id:main, :exception=>"Java::JavaLang::IllegalStateException", :message=>"Unable to configure plugins: (TypeError) Java type is not serializable, cannot be marshaled class org.logstash.config.ir.imperative.PluginStatement", :backtrace=>["org.logstash.config.ir.CompiledPipeline.<init>(CompiledPipeline.java:120)", "org.logstash.execution.JavaBasePipelineExt.initialize(JavaBasePipelineExt.java:85)", "org.logstash.execution.JavaBasePipelineExt$INVOKER$i$1$0$initialize.call(JavaBasePipelineExt$INVOKER$i$1$0$initialize.gen)", "org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:846)", "org.jruby.ir.runtime.IRRuntimeHelpers.instanceSuper(IRRuntimeHelpers.java:1229)", "org.jruby.ir.instructions.InstanceSuperInstr.interpret(InstanceSuperInstr.java:131)", "org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:361)", "org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:72)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:128)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:115)", "org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:329)", "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:87)", "org.jruby.RubyClass.newInstance(RubyClass.java:909)", "org.jruby.RubyClass$INVOKER$i$newInstance.call(RubyClass$INVOKER$i$newInstance.gen)", "org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:329)", "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:87)", "org.jruby.ir.instructions.CallBase.interpret(CallBase.java:549)", "org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:361)", "org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:72)", "org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:92)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:238)", "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:225)", "org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:226)", "org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:393)", "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:206)", "org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:325)", "org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:72)", "org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(Interpreter.java:116)", "org.jruby.runtime.MixedModeIRBlockBody.commonYieldPath(MixedModeIRBlockBody.java:136)", "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:66)", "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)", "org.jruby.runtime.Block.call(Block.java:143)", "org.jruby.RubyProc.call(RubyProc.java:309)", "org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:107)", "java.base/java.lang.Thread.run(Thread.java:833)"]}
```

giving nothing actionable for the user to do to fix their issue

### PR DESCRIPTION
**Pull request by edmocosta**
## Release notes

Fixed pipeline compilation to make it fail when multiple codecs per plugin are specified

## What does this PR do?

Logstah currently does not support multiple codecs per input and output plugins. When the user erroneously configures a plugin with multiple codecs, Logstah behaves differently depending on the configuration.

If the plugin configuration has more than one codec specified, but none of them are setting an attribute/configuration, Logstash starts normally (with no warning/errors messages) and defaults to the first codec, For example, the following configuration would use the `plain` codec and ignore the `json` one:

```
input {
  generator {
    count => 2
    message => '{"id": 1}'
    codec => plain
    codec => json
  }
}
```

On the other hand, if we set any codec's attribute, in this example the `charset`, it fails to start and results in a serialization error (#14773).

```
input {
  generator {
    count => 2
    message => '{"id": 1}'
    codec => plain {
        charset => 'UTF-8'
    }
    codec => json
  }
}
```
Plugins with multiple codecs may result in additional unexpected behavior, such as pipeline or plugin crashes.
This PR fixes the pipeline compilation, throwing a configuration error when the AST is built, failing fast, and avoiding Logstash starting with an invalid configuration.

## Why is it important/What is the impact to the user?

This PR fixes the pipeline compilation ensuring that only one codec per plugin is specified on the configuration.
The current behavior of silently picking the first codec might give users the impression that the configuration is valid and that both codecs are being applied, which is not true.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files (and/or docker env variables)~~
- [x] I have added tests that prove my fix is effective or that my feature works

## How to test this PR locally

Start a process with a pipeline containing multiple codecs configured for the same input or output plugin, for example:

```
input {
  stdin {
    codec => json
    codec => plain 
  }
}

output {
  stdout { codec => rubydebug }
}
```
Observe the configuration error on the logs.

## Related issues

- Closes #14773

## Logs

Erroneously configured pipeline with multiple codecs:

```
input {
  stdin {
    codec => plain{
      charset => "UTF-8"
    } 
    codec => plain 
  }
}

output {
  stdout { codec => rubydebug }
}
```

Error message **before** this change:

```
[2022-12-05T14:12:52,144][ERROR][logstash.agent           ] Failed to execute action {:action=>LogStash::PipelineAction::Create/pipeline_id:main, :exception=>"Java::JavaLang::IllegalStateException", :message=>"Unable to configure plugins: (TypeError) Java type is not serializable, cannot be marshaled class org.logstash.config.ir.imperative.PluginStatement", :backtrace=>["org.logstash.config.ir.CompiledPipeline.<init>(CompiledPipeline.java:120)", "org.logstash.execution.AbstractPipelineExt.initialize(AbstractPipelineExt.java:181)", "org.logstash.execution.AbstractPipelineExt$INVOKER$i$initialize.call(AbstractPipelineExt$INVOKER$i$initialize.gen)", "org.jruby.dist/org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:846)", "org.jruby.dist/org.jruby.ir.runtime.IRRuntimeHelpers.instanceSuper(IRRuntimeHelpers.java:1229)", "org.jruby.dist/org.jruby.ir.instructions.InstanceSuperInstr.interpret(InstanceSuperInstr.java:131)", "org.jruby.dist/org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:361)", "org.jruby.dist/org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:72)", "org.jruby.dist/org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:128)", "org.jruby.dist/org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:115)", "org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:329)", "org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:87)", "org.jruby.dist/org.jruby.RubyClass.newInstance(RubyClass.java:911)", "org.jruby.dist/org.jruby.RubyClass$INVOKER$i$newInstance.call(RubyClass$INVOKER$i$newInstance.gen)", "org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:329)", "org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:87)", "org.jruby.dist/org.jruby.ir.instructions.CallBase.interpret(CallBase.java:549)", "org.jruby.dist/org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:361)", "org.jruby.dist/org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:72)", "org.jruby.dist/org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:92)", "org.jruby.dist/org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:238)", "org.jruby.dist/org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:225)", "org.jruby.dist/org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:226)", "org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:393)", "org.jruby.dist/org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:206)", "org.jruby.dist/org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:325)", "org.jruby.dist/org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:72)", "org.jruby.dist/org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(Interpreter.java:116)", "org.jruby.dist/org.jruby.runtime.MixedModeIRBlockBody.commonYieldPath(MixedModeIRBlockBody.java:136)", "org.jruby.dist/org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:66)", "org.jruby.dist/org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)", "org.jruby.dist/org.jruby.runtime.Block.call(Block.java:143)", "org.jruby.dist/org.jruby.RubyProc.call(RubyProc.java:309)", "org.jruby.dist/org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:107)", "java.base/java.lang.Thread.run(Thread.java:833)"]}
```

Error message **after** this change:
```
[2022-12-05T14:08:16,452][ERROR][logstash.agent           ] Failed to execute action {:action=>LogStash::PipelineAction::Create/pipeline_id:main, :exception=>"LogStash::ConfigurationError", :message=>"Multiple codecs found in the stdin input plugin configuration:\nAt line: 2, column 3", :backtrace=>[...]}

```

### CODE REVIEW

**Comment by edmocosta** [comment]
Jenkins, please test this.

**Comment by edmocosta** [comment]
Jenkins, please test this

**Comment by edmocosta** [comment]
jenkins please test this

**Review by andsel — APPROVED** [review]
LGTM :clap:

File: logstash-core/locales/en.yml [hunk 1]
```diff
@@ -188,6 +188,8 @@ en:
           Duplicate keys found in your configuration: [%{keys}]
           At line: %{line}, column %{column} (byte %{byte})
           after %{after}
+        invalid_plugin_settings_multiple_codecs: |-
+          The %{type} plugin %{plugin} at line %{line} has multiple codecs configured - this is not a valid configuration, only one codec per plugin is supported.
```
Comment (yaauie) [inline logstash-core/locales/en.yml:192]: We do have plugins that support multiple codecs (such as the http input's `additional_codecs` map), so the following is a slight re-wording to make it clear that we are talking about the top-level `codec` option, not _codecs_ as an abstract concept.

```suggestion
          The %{type} plugin %{plugin} at line %{line} has multiple `codec` declarations - this is not a valid configuration, only one top-level `codec` per %{type} plugin is supported.
```

Comment (edmocosta) [reply logstash-core/locales/en.yml:192]: That's great! thank you!

**Pull request merged**